### ISSUE TITLE
Bad accuracy of ttnn.transformer.scaled_dot_product_attention

### ISSUE THREAD

**Issue by philei-tt** [tenstorrent/tt-metal#20710]
### Describe the bug

While trying to remove fallback for flash attention in ttnn compiler (https://github.com/tenstorrent/pytorch2.0_ttnn/pull/569) I noticed that ttnn.transformer.scaled_dot_product_attention has up to 0.96 pcc for some shapes. 

### Steps to reproduce the issue

I used latest commit at the moment of creating this issue: fc1f2d3059f159a61183f1d72ca9a1e86a3b08bb

<details>
  <summary>Here is a code snippet (spoiler) </summary>
  
```
import torch
import ttnn

device_id = 0
dispatch_core_type = ttnn.device.DispatchCoreType.ETH
device = ttnn.open_device(
    device_id=device_id, l1_small_size=8192, dispatch_core_config=ttnn.device.DispatchCoreConfig(dispatch_core_type)
)
ttnn.enable_program_cache(device)


input_shapes = [
    (1, 16, 197, 64),
    (1, 12, 197, 64),
    (1, 16, 50, 64),
    # (1, 8, 4096, 40),
    # (1, 8, 1024, 80),
    (1, 8, 256, 160),
    (1, 8, 64, 160),
    (1, 12, 50, 64),
    # (1, 16, 1370, 80),
    (1, 12, 1, 64),
    (1, 12, 4, 64),
]


def test_attention():
    torch.manual_seed(0)
    for input_shape in input_shapes:
        query = torch.rand(input_shape, dtype=torch.bfloat16)
        key = torch.rand(input_shape, dtype=torch.bfloat16)
        value = torch.rand(input_shape, dtype=torch.bfloat16)

        torch_out = torch.nn.functional.scaled_dot_product_attention(query, key, value, is_causal=False)

        tt_query = ttnn.from_torch(query, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)
        tt_key = ttnn.from_torch(key, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)
        tt_value = ttnn.from_torch(value, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat16, device=device)

        tt_out = ttnn.transformer.scaled_dot_product_attention(
            tt_query,
            tt_key,
            tt_value,
            is_causal=False,
        )

        tt_out_cpu = ttnn.to_torch(tt_out)
        pcc = ttnn.pearson_correlation_coefficient(
            tt_out_cpu,
            torch_out,
        )
        print(f"shape: {input_shape}, pcc: {pcc}")


test_attention()
```
</details>

That produces: 
```
shape: (1, 16, 197, 64), pcc: 0.9642244971517028
shape: (1, 12, 197, 64), pcc: 0.962257374807264
shape: (1, 16, 50, 64), pcc: 0.9950405530785038
shape: (1, 8, 256, 160), pcc: 0.9833102699151345
shape: (1, 8, 64, 160), pcc: 0.9967617869817509
shape: (1, 12, 50, 64), pcc: 0.9947547503763848
shape: (1, 12, 1, 64), pcc: 0.9999895708906906
shape: (1, 12, 4, 64), pcc: 0.9993930657672739
```

### Expected behavior

Ideally pcc should be above 0.999

### Please complete the following environment information

- OS version: Ubuntu 24.04
- python version: 3.10.17
- commit: fc1f2d3059f159a61183f1d72ca9a1e86a3b08bb

**Comment by philei-tt** [comment]
@ayerofieiev-tt @bbradelTT Please take a look

**Comment by bbradelTT** [comment]
You need to use higher fidelity compute kernels if you want better PCC.

Using HiFi4 with non-approx mode and without l1 acc
```
        compute_kernel_config = ttnn.WormholeComputeKernelConfig(
            math_fidelity=ttnn.MathFidelity.HiFi4,
            math_approx_mode=False,
            fp32_dest_acc_en=False,
            packer_l1_acc=False,
        )
        tt_out = ttnn.transformer.scaled_dot_product_attention(
            tt_query,
            tt_key,
            tt_value,
            is_causal=False,
            compute_kernel_config=compute_kernel_config,
        )
```
leads to
```
shape: (1, 16, 197, 64), pcc: 0.9887501121876109
shape: (1, 12, 197, 64), pcc: 0.988533643279507
shape: (1, 16, 50, 64), pcc: 0.9977332211332046
shape: (1, 8, 256, 160), pcc: 0.9865355952572187
shape: (1, 8, 64, 160), pcc: 0.9973754126481726
shape: (1, 12, 50, 64), pcc: 0.9976511831813679
shape: (1, 12, 1, 64), pcc: 0.9999875433531281
shape: (1, 12, 4, 64), pcc: 0.999666400509682
```

Using fp32_dest_acc_en would lead to higher PCCs, but then you'd also need to adjust stack sizes:
```
shape: (1, 16, 197, 64), pcc: 0.990314328594285
shape: (1, 12, 197, 64), pcc: 0.9900980958211115
shape: (1, 16, 50, 64), pcc: 0.9984496499792035
                 Always | FATAL    | Watcher detected stack overflow on Device 0 Core (x=18,y=18): trisc0! Kernel ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/sdpa.cpp uses 320/320 of the stack.
```

@cglagovichTT is there any info about expected PCCs for SDPA?

**Comment by philei-tt** [comment]
Thank you for detailed explanation, I will try that and hope this will be good enough for overall model accuracy. I see that sometimes for lower fidelity we get 0.999+ pcc. Is there some heuristic to know when should we use higher fidelity?

**Comment by bbradelTT** [comment]
@philei-tt fidelity and perf are tradeoffs. If you start with fidelity, try to use the highest fidelity that you are okay with in terms of perf. If you start with perf, try to use the lowest fidelity that you are okay with in terms of PCC.

**Comment by cglagovichTT** [comment]
All tests in `tests/tt_eager/python_api_testing/unit_testing/misc/test_scaled_dot_product_attention.py` get at least 0.99 PCC.

Please check your SDPA program config, trying to make the K chunk size as large a power of 2 as possible such that you don't run out of L1. In addition ensure to set the last argument to false for good accuracy:
```
    program_config = ttnn.SDPAProgramConfig(
        compute_with_storage_grid_size=device.compute_with_storage_grid_size(),
        q_chunk_size=q_chunk_size,
        k_chunk_size=k_chunk_size,
        exp_approx_mode=False, # <- always set false
    )
```
